src/web.rs

```diff
@@ -138,17 +140,62 @@ impl WebImpl {
     }
 
     fn present_with_damage(&mut self, damage: &[Rect]) -> Result<(), SoftBufferError> {
-        let (width, _height) = self
+        let (buffer_width, _buffer_height) = self
             .size
             .expect("Must set size of surface before calling `present_with_damage()`");
+
+        let mut damage_iter = damage.iter();
+
+        let first_rect = if let Some(rect) = damage_iter.next() {
+            rect
+        } else {
+            // If there is no damage, there is nothing to do
+            return Ok(());
+        };
+
+        struct UnionRegion {
+            top: u32,
+            left: u32,
+            bottom: u32,
+            right: u32,
+        }
+
+        let union_region = UnionRegion {
+            top: first_rect.y,
+            left: first_rect.x,
+            bottom: (first_rect.y + first_rect.height.get()),
+            right: (first_rect.x + first_rect.width.get()),
+        };
+
+        let union_region = damage_iter.fold(union_region, |mut union, rect| {
+            union.top = cmp::min(union.top, rect.y);
+            union.left = cmp::min(union.left, rect.x);
+            union.bottom = cmp::max(union.bottom, rect.y + rect.height.get());
+            union.right = cmp::max(union.right, rect.x + rect.width.get());
+            union
+        });
```

Comment:
You should use `reduce` instead of `fold` here.

Reply:
I agree this is almost perfect for `reduce`, but the accumulator is not a `Rect` so I can’t exactly do that

Reply:
I used a `Iterator::map()` to do the trick.

Reply:
Nice, this is actually much better